futures-util/src/io/buf_reader.rs

```diff
@@ -0,0 +1,259 @@
+use futures_core::task::{Context, Poll};
+use futures_io::{AsyncBufRead, AsyncRead, AsyncSeek, Initializer, IoVec, SeekFrom};
+use pin_utils::{unsafe_pinned, unsafe_unpinned};
+use std::io::{self, Read};
+use std::pin::Pin;
+use std::{cmp, fmt};
+use super::DEFAULT_BUF_SIZE;
+
+/// The `BufReader` struct adds buffering to any reader.
```

Comment:
`BufReader` vs `AsyncBufReader`

(I have no strong opinion.)

Reply:
Given it's a concrete type, and not a trait, I think `BufReader` is fine. It's similar to the naming scheme in `Romio`, `Tokio`, and `Runtime` :+1:

Reply:
I wonder if `AsyncRead` and `AsyncWrite` should've been named `PollRead` and `PollWrite` (because they only have polling read/write traits).

Then we could rename `AsyncReadExt` and `AsyncWriteExt` to simply `Read` and `Write`.

I don't see much point in the `Async` prefix if literally no other trait or type has it. Also, the  `Async*` traits with polling methods are not indented to be used by end-users so it feels wrong for them to have simpler names than the `*Ext` traits which are used more often.

Reply:
@stjepang that's an interesting point; perhaps opening a separate issue might be good for a discussion?

Reply:
@stjepang the `Async*` traits are what should be used in bounds so likely used multiple times per file, whereas the `*Ext` traits should be imported and then be used via method syntax so only mentioned once per file.

Reply:
Most relevant issue: https://github.com/rust-lang-nursery/futures-rs/issues/1398